pkg/roll/update_test.go

```diff
@@ -683,8 +684,88 @@ func TestShouldRollMidwayHealthyMigrationFromZeroWhenNewSatisfies(t *testing.T)
 	upd.DesiredReplicas = 3
 	upd.MinimumReplicas = 2
 
-	remove, add, err := upd.shouldRollAfterDelay(rc_fields.RC{ID: upd.NewRC, Manifest: manifest})
+	remove, add, err := upd.shouldRollAfterDelay(manifest.ID())
 	Assert(t).IsNil(err, "expected no error determining nodes to roll")
 	Assert(t).AreEqual(remove, 0, "expected to remove no nodes")
 	Assert(t).AreEqual(add, 1, "expected to add one node")
 }
+
+func panicOnError(desc string, errs <-chan error) {
+	err, ok := <-errs
+	if ok {
+		panic("Error received on " + desc + ":" + err.Error())
+	}
+}
+
+// Transfers the named node from the old RC to the new RC
+func transferNode(node string, manifest pods.Manifest, upd update) error {
+	if _, err := upd.kps.SetPod(kp.REALITY_TREE, node, manifest); err != nil {
+		return err
+	}
+	podID := manifest.ID().String()
+	return upd.labeler.SetLabel(labels.POD, node+"/"+podID, rc.RCIDLabel, string(upd.NewRC))
+}
+
+func TestRollLoopTypicalCase(t *testing.T) {
+	upd, _, manifest := updateWithHealth(t, 3, 0, nil, nil, nil)
+	upd.DesiredReplicas = 3
+	upd.MinimumReplicas = 2
+
+	updates := make(chan map[string]health.Result)
+	errors := make(chan error)
+	quit := make(chan struct{})
+
+	oldRC := rc_fields.RC{ID: upd.OldRC}
+	oldRCQuit := make(chan struct{})
+	oldRCUpdated, oldRCErrors := upd.rcs.Watch(&oldRC, oldRCQuit)
+	go panicOnError("old RC", oldRCErrors)
+
+	newRC := rc_fields.RC{ID: upd.NewRC}
+	newRCQuit := make(chan struct{})
+	newRCUpdated, newRCErrors := upd.rcs.Watch(&newRC, newRCQuit)
+	go panicOnError("new RC", newRCErrors)
+
+	rollLoopResult := make(chan bool)
+
+	go func() {
+		rollLoopResult <- upd.rollLoop(manifest.ID(), updates, errors, quit)
+		close(rollLoopResult)
+	}()
+
+	go panicOnError("roll loop", errors)
```

Comment:
this call incorrectly assumes that `rollLoop` *sends* errors on the error channel. Nope, it doesn't. `rollLoop` expects to *receive* errors on this channel, so again we can just pass nil.